### DISCUSSION TITLE
How to upgrade all packages installed by winget only?

### DISCUSSION THREAD

**Discussion by nkrepo** [microsoft/winget-cli#3290] [General] [answered]
Since many users are using multiple package managers like choco e.t.c and also manually installe packages that they don't want to be updated automatically by `winget upgrade -h --all`
Is there an easy way to accomplish that behavior or we have to keep track manually the packages installed by winget ?

**Comment by mdanish-kh** [comment, marked as answer]
WinGet doesn't have a way to track whether a package was installed through it or any external means

To achieve your desired behavior, take a look at [Package Pinning](https://github.com/microsoft/winget-cli/blob/master/doc/specs/%23476%20-%20Package%20Pinning.md) feature. You can use `winget pin` (currently available in the [preview release](https://aka.ms/getwingetpreview) of WinGet) for packages that you don't want WinGet to upgrade.

**Reply by nkrepo** [reply]
> WinGet doesn't have a way to track whether a package was installed through it or any external means
> 
> To achieve your desired behavior, take a look at [Package Pinning](https://github.com/microsoft/winget-cli/blob/master/doc/specs/%23476%20-%20Package%20Pinning.md) feature. You can use `winget pin` (currently available in the [preview release](https://aka.ms/getwingetpreview) of WinGet) for packages that you don't want WinGet to upgrade.

Unfortunately that does help, I still would have to manually maintain a list of the packages that are installed by winget. I want to avoid that altogether.

**Reply by Masamune3210** [reply]
Thats the thing, unlike choco or snappy, packages that are installed by winget are installed like any other program. Winget just facilitates the downloading and installation of the installation package. Outside of zip's and other archives once they are added, winget's involvement with the program begins at download and ends after the installation is complete.

**Reply by Jaykul** [reply]
It's not quite true that winget downloads an "installation package" and then has nothing to do with it. Sometimes it also make a path symlink. 

The problem is that "upgrading" a tool that was installed some other way almost never works the way you expect. For example, a huge number of the things available via winget are single binary tools written in go or rust, and don't have "installers." If, you install (just for example) `kubeconform` with winget, you end up with `~\AppData\Local\Microsoft\WinGet\Links\kubeconform.exe` on your PATH, which is a pointer to `~\AppData\Local\Microsoft\WinGet\Packages\YannHamon.kubeconform_Microsoft.Winget.Source_8wekyb3d8bbwe\kubeconform.exe`

The problem is that every other install tool has been doing this exact same thing for years. So if you've installed that tool with `choco` or `scoop` (or by just downloading the executable and putting it in your path) and it's out of date, and you run `winget upgrade` _before_ you run `choco upgrade` ... winget will see it, determine it's out of date, download the latest version and put that symlink in there ...  but _most likely_ the version that you _already_ had installed is still the one that shows up first in the path ...

So now you have two copies. _You are still using the outdated one_. Winget will no longer show that it's outdated -- because it sees both, and does not know/care that it's the old one is the one that's being used.

**Reply by Masamune3210** [reply]
All of this was written before portable application support (the thing you are talking about where there isn't an installer, just a program) was added
